core/src/bank_forks.rs

```diff
@@ -211,31 +333,65 @@ impl BankForks {
             .map_err(|_| BankForks::get_io_error("serialize bank status cache error"))?;
         serialize_into(&mut stream, &bank.rc)
             .map_err(|_| BankForks::get_io_error("serialize bank accounts error"))?;
+
+        // Create hard links to each AppendVec that's necessary to restore from this
+        // snapshot, so that these AppendVec's don't get removed.
+        let account_storage_entries = bank.rc.get_storage_entries();
+        for s in account_storage_entries {
+            let append_vec_file_path = s.get_path();
+            let append_vec_dir_path = Path::new(
+                append_vec_file_path
+                    .parent()
+                    .expect("Invalid AppendVec file path"),
+            );
+            let link = append_vec_dir_path.join(Self::get_snapshot_appendvec_hardlink_name(
+                slot,
+                s.fork_id(),
+                s.append_vec_id(),
+            ));
+            fs::hard_link(&append_vec_file_path, &link)?;
```

Comment:
@rwalker, here's where I'm adding the hard links to the AppendVecs when generating the snapshot. I still need to add the logic to remove these in `remove_snapshot()`. We could deserialize the snapshot to figure out which AppendVecs have hard links we need to delete, or we could stash that information elsewhere. Thoughts?